common/src/utils.rs

```diff
@@ -1,49 +1,153 @@
+pub const INJECTIVE_BYTES_PER_ELEMENT: usize = 4;
+pub const DIGEST_BYTES_PER_ELEMENT: usize = 8;
+pub const FELTS_PER_U128: usize = 4;
+pub const FELTS_PER_U64: usize = 2;
 pub const DIGEST_NUM_FIELD_ELEMENTS: usize = 4;
 
-pub const ZERO_DIGEST: Digest = [F::ZERO; 4];
+pub const ZERO_DIGEST: Digest = [F::ZERO; DIGEST_NUM_FIELD_ELEMENTS];
+pub const BIT_32_LIMB_MASK: u64 = 0xFFFF_FFFF;
 
-pub type Digest = [F; 4];
+pub type Digest = [F; DIGEST_NUM_FIELD_ELEMENTS];
 pub type PrivateKey = [F; 4];
 
+#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
+pub struct BytesDigest(pub [u8; 32]);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum DigestError {
+    ChunkOutOfFieldRange { chunk_index: usize, value: u64 },
+    InvalidLength { expected: usize, got: usize },
+}
+
+impl TryFrom<&[u8]> for BytesDigest {
+    type Error = DigestError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        let bytes: [u8; 32] = value.try_into().map_err(|_| DigestError::InvalidLength {
+            expected: 32,
+            got: value.len(),
+        })?;
+        for (i, chunk) in bytes.chunks(8).enumerate() {
+            let v = u64::from_le_bytes(chunk.try_into().unwrap());
+            if v >= F::ORDER {
+                return Err(DigestError::ChunkOutOfFieldRange {
+                    chunk_index: i,
+                    value: v,
+                });
+            }
+        }
+        Ok(BytesDigest(bytes))
+    }
+}
+
+impl From<[u8; 32]> for BytesDigest {
+    fn from(value: [u8; 32]) -> Self {
+        for (i, chunk) in value.chunks(8).enumerate() {
+            let v = u64::from_le_bytes(chunk.try_into().unwrap());
+            if v >= F::ORDER {
+                panic!("Invalid digest value: chunk {} out of range: {}", i, v);
+            }
+        }
+        BytesDigest(value)
+    }
+}
+
+impl From<Digest> for BytesDigest {
+    fn from(value: Digest) -> Self {
+        let bytes = digest_felts_to_bytes(value);
+        Self(*bytes)
+    }
+}
+
+impl TryFrom<&[F]> for BytesDigest {
+    type Error = anyhow::Error;
+
+    fn try_from(value: &[F]) -> Result<Self, Self::Error> {
+        let digest: Digest = value.try_into().map_err(|_| {
+            anyhow!(
+                "failed to deserialize bytes digest from field elements. Expected length 4, got {}",
+                value.len()
+            )
+        })?;
+        let bytes = digest_felts_to_bytes(digest);
+        Ok(Self(*bytes))
+    }
+}
+
+impl Deref for BytesDigest {
+    type Target = [u8; 32];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
 pub fn u128_to_felts(num: u128) -> [F; FELTS_PER_U128] {
-    let amount_high = F::from_noncanonical_u64((num >> 64) as u64);
-    let amount_low = F::from_noncanonical_u64(num as u64);
-    [amount_high, amount_low]
+    (0..FELTS_PER_U128)
+        .map(|i| {
+            let shift = 96 - 32 * i;
+            let limb = ((num >> shift) & BIT_32_LIMB_MASK as u128) as u64;
```

Comment:
ok so this is putting 32bit limbs into a u64, which is always canonical because F::ORDER > u32::MAX. Can we put a comment in here?